File: core/src/main/java/pcb/uwu/units/composite/termodynamics/EnergyUnit.java
package pcb.uwu.units.composite.termodynamics;

import pcb.uwu.core.CompositeUnit;
import pcb.uwu.core.UnitCounter;
import pcb.uwu.units.composite.electromagnetism.ElectricChargeUnit;
import pcb.uwu.units.composite.electromagnetism.ElectricPotentialUnit;
import pcb.uwu.units.composite.fundamental.VolumeUnit;
import pcb.uwu.units.composite.mechanics.ForceUnit;
import pcb.uwu.units.quantity.LengthUnit;
import pcb.uwu.units.quantity.MassUnit;
import pcb.uwu.units.quantity.TimeUnit;

public class EnergyUnit extends CompositeUnit {

	// region base constructor

	public EnergyUnit(LengthUnit lengthUnit, MassUnit massUnit, TimeUnit timeUnit) {
		super(new UnitCounter()
				.major(massUnit)
				.major(lengthUnit, 2)
				.minor(timeUnit, 2));
	}

	// endregion

	// region composite constructors

	public EnergyUnit(ForceUnit forceUnit, LengthUnit lengthUnit) {
		super(new UnitCounter()
				.major(forceUnit.getUnitCounter())
				.major(lengthUnit));
	}

	public EnergyUnit(PressureUnit pressureUnit, VolumeUnit volumeUnit) {
		super(new UnitCounter()
				.major(pressureUnit.getUnitCounter())
				.major(volumeUnit.getUnitCounter()));
	}

	public EnergyUnit(PowerUnit powerUnit, TimeUnit timeUnit) {
		super(new UnitCounter()
				.major(powerUnit.getUnitCounter())
				.major(timeUnit));
	}

	public EnergyUnit(ElectricChargeUnit electricChargeUnit, ElectricPotentialUnit electricPotentialUnit) {
		super(new UnitCounter()
				.major(electricChargeUnit.getUnitCounter())
				.major(electricPotentialUnit.getUnitCounter()));
	}

	// endregion

}

File: core/src/main/java/pcb/uwu/unit/derived/termodynamics/EnergyUnit.kt
package pcb.uwu.unit.derived.termodynamics

import pcb.uwu.core.CompositeUnit
import pcb.uwu.core.UnitCounter
import pcb.uwu.unit.base.LengthUnit
import pcb.uwu.unit.base.MassUnit
import pcb.uwu.unit.base.TimeUnit
import pcb.uwu.unit.derived.electromagnetism.ElectricChargeUnit
import pcb.uwu.unit.derived.electromagnetism.ElectricPotentialUnit
import pcb.uwu.unit.derived.fundamental.VolumeUnit
import pcb.uwu.unit.derived.mechanics.ForceUnit

open class EnergyUnit : CompositeUnit
{
    constructor(lengthUnit: LengthUnit,
                massUnit: MassUnit,
                timeUnit: TimeUnit)
            : super(UnitCounter()
                        .major(massUnit)
                        .major(lengthUnit)
                        .major(lengthUnit)
                        .minor(timeUnit)
                        .minor(timeUnit))

    constructor(forceUnit: ForceUnit,
                lengthUnit: LengthUnit)
            : super(UnitCounter()
                        .major(forceUnit)
                        .major(lengthUnit))

    constructor(pressureUnit: PressureUnit,
                volumeUnit: VolumeUnit)
            : super(UnitCounter()
                        .major(pressureUnit)
                        .major(volumeUnit))

    constructor(powerUnit: PowerUnit,
                timeUnit: TimeUnit)
            : super(UnitCounter()
                        .major(powerUnit)
                        .major(timeUnit))

    constructor(electricChargeUnit: ElectricChargeUnit,
                electricPotentialUnit: ElectricPotentialUnit)
            : super(UnitCounter()
                        .major(electricChargeUnit)
                        .major(electricPotentialUnit))
}